File: daily/2024/145_Binary_Tree_Postorder_Traversal.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

TreeNode* create_tree(vector<int>& nodes){
    if (nodes.empty()) return nullptr;
    TreeNode* root = new TreeNode(nodes[0]);
    queue<TreeNode*> que;
    que.push(root);
    int i = 1;
    int n = nodes.size();
    while(i < n && !que.empty()){
        TreeNode* temp = que.front();
        que.pop();
        if (i < n && nodes[i] != -1){
            temp->left = new TreeNode(nodes[i]);
            que.push(temp->left);
        }
        i++;
        if (i < n && nodes[i] != -1){
            temp->right = new TreeNode(nodes[i]);
            que.push(temp->right);
        }
        i++;
    }
    return root;
}

class Solution {
public:
    void traversal(TreeNode* root, vector<int>& postorder){
        if (!root) return;
        traversal(root->left, postorder);
        traversal(root->right, postorder);
        postorder.push_back(root->val);
        return;
    }
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> postorder;
        traversal(root, postorder);
        return postorder;
    }
};

int main(){
    vector<int> nodes = {1,-1,2,3};
    // vector<int> nodes = {};
    // vector<int> nodes = {1};
    TreeNode* root = create_tree(nodes);
    Solution S;
    vector<int> ans = S.postorderTraversal(root);
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}

File: daily_python/145_Binary_Tree_Postorder_Traversal.py
import os
import sys
from typing import List, Optional
import queue

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def create_tree(nodes: List[int])-> TreeNode:
    if not nodes:
        return None
    root = TreeNode(nodes[0])
    que = queue.Queue()
    que.put(root)
    index = 1
    n = len(nodes)
    while index < n and not que.empty():
        temp = que.get()
        if index < n and nodes[index] != -1:
            temp.left = TreeNode(nodes[index])
            que.put(temp.left)
        index += 1
        if index < n and nodes[index] != -1:
            temp.right = TreeNode(nodes[index])
            que.put(temp.right)
        index += 1
    return root

class Solution:
    def traversal(self, root: TreeNode)-> None:
        if not root:
            return
        self.traversal(root.left)
        self.traversal(root.right)
        self.postorder.append(root.val)

    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        self.postorder = []
        self.traversal(root)
        return self.postorder

def readDataSet(filename: str):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nodes = lines[0].split('=')[1].strip()[1:-2]
            if nodes:
                nodes = nodes.split(',')
                data = [int(i) for i in nodes]
            else:
                data = []
            dataset.append(data)
    return dataset

if __name__ == '__main__':
    if (len(sys.argv) == 1):
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nodes in dataset:
        results.append(solution.postorderTraversal(create_tree(nodes)))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')
